File: mods/industrial_era/lua/entities/technology.lua
-- Industrial_Era_mod - Base game files for Symphony of Empires
-- Copyright (C) 2021, Symphony of Empires contributors
-- 
-- This program is free software; you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation; either version 2 of the License, or
-- (at your option) any later version.
-- 
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License along
-- with this program; if not, write to the Free Software Foundation, Inc.,
-- 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
--
-- ----------------------------------------------------------------------------
--	Name:
--  	technology.lua
--
-- 	Abstract:
--      Does important stuff
-- ----------------------------------------------------------------------------

-- Mechanical theory series
mechanical_theory_i = Technology:new{
    ref_name = "mechanical_theory_i",
    name = "Mechanical theory I",
    description = "Relying on natural sources such as the wind, they will be the stepping stone for future developments",
    cost = 1000.0,
    type = TECH_ECONOMIC
}
mechanical_theory_i:register()

mechanical_theory_ii = Technology:new{
    ref_name = "mechanical_theory_ii",
    name = "Mechanical theory II",
    description = "The boom in science will help more complex machinery to rise",
    cost = 180000.0,
    type = TECH_ECONOMIC
}
mechanical_theory_ii:register()
mechanical_theory_ii:requires_technology(mechanical_theory_i)

mechanical_theory_iii = Technology:new{
    ref_name = "mechanical_theory_iii",
    name = "Mechanical theory III",
    description = "With the use of power generators it's possible to automate the movement of machines",
    cost = 180000.0,
    type = TECH_ECONOMIC
}
mechanical_theory_iii:register()
mechanical_theory_iii:requires_technology(mechanical_theory_ii)

voltaic_pile = Technology:new{
    ref_name = "voltaic_pile",
    name = _("Voltaic pile"),
    description = "An early battery",
    cost = 15000.0,
    type = TECH_ECONOMIC
}
voltaic_pile:register()

combustion_engine = Technology:new{
    ref_name = "combustion_engine",
    name = "Combustion engine",
    description = "",
    cost = 50000.0,
    type = TECH_ECONOMIC
}
combustion_engine:register()

steam_locomotive = Technology:new{
    ref_name = "steam_locomotive",
    name = "Steam locomotive",
    description = "",
    cost = 100000.0,
    type = TECH_ECONOMIC
}
steam_locomotive:register()

assembly_line = Technology:new{
    ref_name = "assembly_line",
    name = "Assembly line",
    description = "",
    cost = 100000.0,
    type = TECH_ECONOMIC
}
assembly_line:register()
assembly_line:requires_technology(mechanical_theory_iii)

napoleonic_tought = Technology:new{
    ref_name = "napoleonic_tought",
    name = "Napoleonic tought",
    description = "",
    cost = 10000.0,
    type = TECH_STRATEGIC
}
napoleonic_tought:register()

trench_warfare = Technology:new{
    ref_name = "trench_warfare",
    name = "Trench warfare",
    description = "Develop efficient tactics to create defensive frontlines making offensives very costly",
    cost = 150000.0,
    type = TECH_STRATEGIC
}
trench_warfare:register()

guerrilla_warfare = Technology:new{
    ref_name = "guerrilla_warfare",
    name = "Guerrilla warfare",
    description = "Splitting the army to allow quick-attacks and small raids that weaken and distract units and slow down their movement",
    cost = 150000.0,
    type = TECH_STRATEGIC
}
guerrilla_warfare:register()

-- Social technologies
romanticism = Technology:new{
    ref_name = "romanticism",
    name = "Romanticism",
    description = "...",
    cost = 150000.0,
    type = TECH_SOCIAL
}
romanticism:register()

-- Military and naval tech
gunpowder = Technology:new{
    ref_name = "gunpowder",
    name = "Gunpowder",
    description = "...",
    cost = 150000.0,
    type = TECH_MILITARY
}
gunpowder:register()

flintlock_rifles = Technology:new{
    ref_name = "flintlock_rifles",
    name = "Flintlock Rifles",
    description = "...",
    cost = 150000.0,
    type = TECH_MILITARY
}
flintlock_rifles:register()
flintlock_rifles:requires_technology(gunpowder)

light_vessel = Technology:new{
    ref_name = "light_vessel",
    name = "Light Vessel",
    description = "...",
    cost = 150000.0,
    type = TECH_NAVAL
}
light_vessel:register()

heavy_vessel = Technology:new{
    ref_name = "heavy_vessel",
    name = "Heavy Vessel",
    description = "...",
    cost = 150000.0,
    type = TECH_NAVAL
}
heavy_vessel:register()
heavy_vessel:requires_technology(light_vessel)

File: game/src/client/interface/technology.cpp
// Symphony of Empires
// Copyright (C) 2021, Symphony of Empires contributors
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, write to the Free Software Foundation, Inc.,
// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
//
// ----------------------------------------------------------------------------
// Name:
//      client/interface/technology.cpp
//
// Abstract:
//      Does some important stuff.
// ----------------------------------------------------------------------------

#include "eng3d/path.hpp"
#include "eng3d/texture.hpp"
#include "eng3d/ui/image.hpp"
#include "eng3d/ui/close_button.hpp"
#include "eng3d/ui/button.hpp"

#include "client/interface/technology.hpp"
#include "client/map.hpp"
#include "client/game_state.hpp"
#include "nation.hpp"
#include "world.hpp"
#include "technology.hpp"

using namespace Interface;

TechTreeTab::TechTreeTab(GameState& _gs, int x, int y, int type, UI::Widget* parent)
    : UI::Group(x, y, parent->width - x, parent->height - y, parent),
    gs{ _gs }
{
    unsigned int i = 0;
    for(auto& technology : gs.world->technologies) {
        if(static_cast<int>(technology.type) & static_cast<int>(type)) {
            auto* btn = new TechnologyInfo(gs, 0, 48 * i, technology, this);
            btn->text(technology.ref_name.get_string());
            i++;
        }
    }
}

TechTreeView::TechTreeView(GameState& _gs)
    : UI::Window(-400, 0, 400, _gs.height),
    gs{ _gs }
{
    if(gs.right_side_panel != nullptr) {
        gs.right_side_panel->kill();
    }
    gs.right_side_panel = this;
    this->set_close_btn_function([this](Widget&) {
        this->kill();
        this->gs.right_side_panel = nullptr;
    });

    this->origin = UI::Origin::UPPER_RIGHT_SCREEN;
    this->is_scroll = false;
    this->text("Research");

    this->strategic_tab = new TechTreeTab(gs, 0, 24, TechnologyType::STRATEGIC, this);
    this->strategic_tab->is_render = true;
    auto* strategic_ibtn = new UI::Image(0, 0, 32, 32, gs.tex_man->load(Path::get("gfx/strategic_tech.png")), this);
    strategic_ibtn->set_on_click([this](UI::Widget&) {
        this->strategic_tab->is_render = true;
        this->military_tab->is_render = false;
        this->navy_tab->is_render = false;
        this->social_tab->is_render = false;
        this->economic_tab->is_render = false;
        this->politics_tab->is_render = false;
    });

    this->military_tab = new TechTreeTab(gs, 0, 24, TechnologyType::MILITARY, this);
    this->military_tab->is_render = false;
    auto* military_ibtn = new UI::Image(0, 0, 32, 32, gs.tex_man->load(Path::get("gfx/military_tech.png")), this);
    military_ibtn->right_side_of(*strategic_ibtn);
    military_ibtn->set_on_click([this](UI::Widget&) {
        this->strategic_tab->is_render = false;
        this->military_tab->is_render = true;
        this->navy_tab->is_render = false;
        this->social_tab->is_render = false;
        this->economic_tab->is_render = false;
        this->politics_tab->is_render = false;
    });

    this->navy_tab = new TechTreeTab(gs, 0, 24, TechnologyType::NAVY, this);
    this->navy_tab->is_render = false;
    auto* navy_ibtn = new UI::Image(0, 0, 32, 32, gs.tex_man->load(Path::get("gfx/navy_tech.png")), this);
    navy_ibtn->right_side_of(*military_ibtn);
    navy_ibtn->set_on_click([this](UI::Widget&) {
        this->strategic_tab->is_render = false;
        this->military_tab->is_render = false;
        this->navy_tab->is_render = true;
        this->social_tab->is_render = false;
        this->economic_tab->is_render = false;
        this->politics_tab->is_render = false;
    });

    this->social_tab = new TechTreeTab(gs, 0, 24, TechnologyType::SOCIAL, this);
    this->social_tab->is_render = false;
    auto* social_ibtn = new UI::Image(0, 0, 32, 32, gs.tex_man->load(Path::get("gfx/social_tech.png")), this);
    social_ibtn->right_side_of(*navy_ibtn);
    social_ibtn->set_on_click([this](UI::Widget&) {
        this->strategic_tab->is_render = false;
        this->military_tab->is_render = false;
        this->navy_tab->is_render = false;
        this->social_tab->is_render = true;
        this->economic_tab->is_render = false;
        this->politics_tab->is_render = false;
    });

    this->economic_tab = new TechTreeTab(gs, 0, 24, TechnologyType::ECONOMIC, this);
    this->economic_tab->is_render = false;
    auto* economic_ibtn = new UI::Image(0, 0, 32, 32, gs.tex_man->load(Path::get("gfx/money.png")), this);
    economic_ibtn->right_side_of(*social_ibtn);
    economic_ibtn->set_on_click([this](UI::Widget&) {
        this->strategic_tab->is_render = false;
        this->military_tab->is_render = false;
        this->navy_tab->is_render = false;
        this->social_tab->is_render = false;
        this->economic_tab->is_render = true;
        this->politics_tab->is_render = false;
    });

    this->politics_tab = new TechTreeTab(gs, 0, 0, TechnologyType::POLITICS, this);
    this->politics_tab->is_render = false;
    auto* politics_ibtn = new UI::Image(0, 0, 32, 32, gs.tex_man->load(Path::get("gfx/politics_tech.png")), this);
    politics_ibtn->right_side_of(*economic_ibtn);
    politics_ibtn->set_on_click([this](UI::Widget&) {
        this->strategic_tab->is_render = false;
        this->military_tab->is_render = false;
        this->navy_tab->is_render = false;
        this->social_tab->is_render = false;
        this->economic_tab->is_render = false;
        this->politics_tab->is_render = true;
    });
}